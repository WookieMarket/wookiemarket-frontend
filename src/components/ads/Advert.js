import React from 'react';
import Fields from '../layout/Fields';
import defaultImage from '../../assets/no_image.jpg';
import '../../css/advert.css';

function Advert({ advert, onImageError }) {
  console.log('Rendering Advert component');
  //const advDate = new Date(advert.createdAt);
  if (!advert) {
    console.error('No advertisement with the specified ID found');
    return <p>The specified advertisement could not be found.</p>;
  }

  return (
    <>
      <div className='productInfo' id='advertOnly'>
        <div className='advert-name'>
          <h2>
            <Fields value={advert.name} />
          </h2>
        </div>
      </div>
      <br />
      <div className='productData'>
        <div className='product-img'>
          <img
            className='productPhoto'
            src={advert.image || defaultImage}
            onError={onImageError}
            alt='imagen del producto en venta'
          ></img>
        </div>
        <br />
        <div className='product_Info'>
          <p className='advert_label'>Description: </p>
          <Fields value={advert.description} />
          {/*<span className='description'> {advert.description}</span>*/}
          <br />
          <br />
          <p>
            Is
            {advert.onSale ? (
              <span id='isSale'> for sale </span>
            ) : (
              <span id='isSale'> purchased </span>
            )}
            this product by:
          </p>
          <div className='price'>
            <p>
              <Fields value={advert.price} />
              {/*{advert.price}*/}{' '}
              <span id='price_coin'>
                {' '}
                <Fields value={advert.coin} />
              </span>
            </p>
          </div>
          <div className='advert_label'>
            <p>Category: </p>
            <p>Category: </p>
            <span className='advert_text'>
              {advert.category &&
                advert.category.map((cat) => (
                  <Fields key={cat} value={cat.toLocaleUpperCase()} />
                ))}
            </span>
            {/*</span><span className='advert_text'>
                {advert.category === undefined
                  ? ''
                  : advert.category.join(', ').toLocaleUpperCase()}
              </span>
            */}
          </div>
        </div>
        <div>
          <p className='advert_label'>
            Username: <Fields value={advert.username} />
            {/*<span className='advert-text'>{advert.username}</span>*/}
          </p>
          <small className='advert_label'>
            Created at:{' '}
            {
              //<span className='advert_text'>{`${advDate.toUTCString()}`}</span>
            }
          </small>
        </div>
      </div>
    </>
  );
}

export default Advert;
