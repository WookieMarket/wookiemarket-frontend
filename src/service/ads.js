import client from '../api/client';

const adUrl = '/api/ads/adverts';

export const getAd = adId => {
  const url = `${adUrl}/${adId}`;
  return client.get(url);
};

export const createAd = ad => {
  const config = {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  };
  const url = `${adUrl}/create`;
  return client.post(url, ad, config);
};

export const modifyAd = (adId, ad) => {
  const config = {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  };
  const url = `${adUrl}/update/${adId}`;
  return client.put(url, ad, config);
};

// export const modifyAd = ad => {
//   const config = {
//     headers: {
//       'Content-Type': 'multipart/form-data',
//     },
//   };
//   const url = `${adUrl}/update`;
//   return client.post(url, ad, config);
// };

export const getTags = () => {
  const url = `${adUrl}/tags`;
  return client.get(url);
};

// export const getLastAdv = () => {
//   const url = `${adUrl}`;
//   return client.get(url);
// };

export const getRecentAds = () => {
  const url = `${adUrl}`;
  return client.get(url);
};
