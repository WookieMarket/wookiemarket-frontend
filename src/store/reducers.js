export {default as adverts} from './slices/adverts';
export {default as ui} from './slices/ui';