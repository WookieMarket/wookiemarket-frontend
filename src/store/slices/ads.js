import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { areAdvertsLoaded } from '../selectors';

export const adsCreate = createAsyncThunk(
  'ads/create',
  async (ad, { extra: { service }, rejectWithValue }) => {
    try {
      //TODO modificar cuando este implementado el detalle del anuncio
      //const { createAd } = await service.ads.createAd(ad);
      const createAd = await service.ads.createAd(ad);
      console.log('Ante', createAd);

      const createdAdId = createAd.result._id;

      //TODO modificar cuando este implementado el detalle del anuncio
      const fetchedAd = await service.ads.getAd(createdAdId);
      //return createdAd;
      return fetchedAd;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const advertsList = createAsyncThunk(
  'ads/list',
  async ({ skip, limit, sort }, { extra: { service }, rejectWithValue }) => {
    try {

      const adverts = await service.ads.getRecentAds(skip, limit);

      return adverts;
    } catch (error) {
      console.log(error);
      return rejectWithValue(error);
    }
  },
  {
    condition: (_, { getState }) => !areAdvertsLoaded(getState()),
  },
);

export const getAdById = createAsyncThunk(
  'ads/fetchAdvertById',
  async (id, { extra: { service }, rejectWithValue }) => {
    try {
      const advert = await service.ads.getAd(id);
      return advert;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const uploadModifiedAd = createAsyncThunk(
  'ads/uploadModifiedAd',
  async ({ id, ad }, { extra: { service }, rejectWithValue }) => {
    try {
      //ad = await service.ads.getAd(id);
      const modifiedAd = await service.ads.modifyAd(id, ad);
      return modifiedAd;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const deleteAdvert = createAsyncThunk(
  'ads/deleteAdvert',
  async (id, { extra: { service }, rejectWithValue }) => {
    try {
      await service.ads.deleteAdvert(id);
      console.log('Advert...delted');
      return true;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const setAdsPerPage = createAsyncThunk(
  'ads/setAdsPerPage',
  async (adsPerPage, { dispatch }) => {
    dispatch({ type: 'ads/setAdsPerPage', payload: adsPerPage });
  }
);

const ads = createSlice({
  name: 'ads',
  initialState: {
    areLoaded: false,
    data: [],
    adsPerPage: 4,
    totalCountAds: 0,
  },
  reducers: {
    setAdsPerPage(state, action) {
      state.adsPerPage = action.payload;
    },
  },
  extraReducers: builder => {
    builder
      .addCase(adsCreate.fulfilled, (state, action) => {
        state.data.unshift(action.payload.result);
      })
      .addCase(advertsList.fulfilled, (state, action) => {
        state.areLoaded = true;
        state.data = action.payload.results;
        state.totalCountAds = action.payload.total;
      })
      .addCase(uploadModifiedAd.fulfilled, (state, action) => {
        //state.areLoaded = true;
        state.data.unshift(action.payload.result);
      })
      .addCase(getAdById.fulfilled, (state, action) => {
        state.areLoaded = false;
        state.data = [action.payload.result];
      });
  },
});

export default ads.reducer;
