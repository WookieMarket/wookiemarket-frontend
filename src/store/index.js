import { configureStore as rtkConfigureStore } from '@reduxjs/toolkit';
import * as service from '../service';
import * as reducers from './reducers';
import { failureRedirects, successRedirects } from './middleware';
import { authSignup, authLogin } from './slices/auth';

export default function configureStore(preloadedState, { router }) {
  const extraMiddleware = [
    failureRedirects(router, { 401: '/login', 404: '/404' }),
    successRedirects(router, {
      [authSignup.fulfilled.type]: () => {
        return router.state.location.state?.from?.pathname || '/';
      },
      [authLogin.fulfilled.type]: () => {
        return router.state.location.state?.from?.pathname || '/';
      },
    }),
  ];

  const store = rtkConfigureStore({
    reducer: reducers,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        thunk: { extraArgument: { service, router } },
        serializableCheck: false,
      }).concat(extraMiddleware),
    preloadedState,
  });
  return store;
}
