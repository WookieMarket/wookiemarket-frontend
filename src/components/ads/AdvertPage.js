import React from 'react';
import '../../css/advert.css';
import Advert from './Advert';
import { useLocation, useNavigate, useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { getAdvertById } from '../../store/selectors';

const AdvertPage = ({ onImageError }) => {
  const { advertId } = useParams();
  /*const location = useLocation();
  const searchParams = new URLSearchParams(location.search);
  const advertId = searchParams.get('_id');*/

  console.log('Advert ID:', advertId);

  const advert = useSelector((state) => getAdvertById(state, advertId));

  console.log('Esto debería ser el anuncio:', advert);
 

  const navigate = useNavigate();
  //navigate.replace(`/adverts/${advert.name}`);

  if (advert) {
    navigate(`/adverts/${advert._id}`);
  } else {
    console.error('No se encontró un anuncio con el ID especificado');
  }

  return (
    <>
      <Advert advert={advert} onImageError={onImageError} />
      <div>
        <small>Created at: ¡AÑADIR CAMPO FECHA CREACIÓN ANUNCIO!</small>
      </div>
    </>
  );
};

export default AdvertPage;
