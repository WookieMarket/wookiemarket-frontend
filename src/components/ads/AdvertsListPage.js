import { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import Advert from "./Advert";
import { useDispatch, useSelector } from "react-redux";
import { getAdverts, getUi } from "../../store/selectors";
import { advertsList } from "../../store/slices/ads";
import "../../css/advertListPage.css";
import "../../css/advert.css";
import defaultImage from "../../assets/no_image.jpg";
import { filterByName } from "../../store/slices/adsFiltered";
import Layout from "../layout/Layout";

const EmptyList = ({ dataFiltered }) => {
  return (
    <div style={{ textAlign: "center" }}>
      <p>Sorry, no adverts yet.</p>
    </div>
  );
};

const handleImageError = event => {
  event.target.src = defaultImage;
};

const advertsPerPage = 2;

const AdvertsListPage = () => {
  //const [query, setQuery] = useState('');
  const [currentPage, setCurrentPage] = useState(1);
  const [filteredAdverts, setFilteredAdverts] = useState("");
  const [isFilterActive, setIsFilterActive] = useState(false);
  const ads = useSelector(getAdverts);
  const { isLoading } = useSelector(getUi);
  const dispatch = useDispatch();

  // useEffect(() => {
  //   onAdvertsLoaded();
  // }, [onAdvertsLoaded]);

  useEffect(() => {
    dispatch(advertsList()).catch(error => console.log(error));
  }, [dispatch]);

  const handlePageChange = page => {
    setCurrentPage(page);
  };

  const filterAdName = ad =>
    (ad.name ?? "").toUpperCase().startsWith(filteredAdverts.toUpperCase());

  const filteredAds = ads
    //.filter(filterAdSaleValueOrDefault)
    .filter(filterAdName);
  //.filter(filterAdPrice)
  //.filter(filterAdTags);

  //const totalPages = Math.ceil(adverts.length / advertsPerPage);
  const totalPages = Math.ceil(
    (isFilterActive ? filteredAds.length : ads.length) / advertsPerPage,
  );
  const startIndex = (currentPage - 1) * advertsPerPage;
  const endIndex = startIndex + advertsPerPage;
  //const advertsToDisplay = adverts.slice(startIndex, endIndex);
  const advertsToDisplay = isFilterActive
    ? filteredAds.slice(startIndex, endIndex)
    : ads.slice(startIndex, endIndex);
  //const isLastPage = currentPage === totalPages;
  const isLastPage = isFilterActive
    ? currentPage === Math.ceil(filteredAds.length / advertsPerPage)
    : currentPage === Math.ceil(ads.length / advertsPerPage);

  // const handleFilterChange = event => {
  //   // //filterByName(event.target.value);
  //   // const filteredData = filterByName({ name: event.target.value, adverts });
  //   // setFilteredAdverts(filteredData);
  //   setFilteredAdverts(event.target.value);
  // };

  // const handleFilterChange = event => {
  //   const value = event.target.value;
  //   setFilteredAdverts(value);
  //   setIsFilterActive(value.trim() !== "");
  // };
  const handleFilterChange = event => {
    const value = event.target.value;
    setFilteredAdverts(value);
    setIsFilterActive(value.trim() !== "");
    dispatch(filterByName({ name: value.trim(), adverts: ads })); // Llama a la acci√≥n filterByName
  };

  return (
    <Layout title="anuncios">
      <>
        <section className="searchSection">
          <h1>Searching area</h1>
          <label className="advert_label">Name: </label>
          <input type="text" onChange={handleFilterChange} />
        </section>
        <div className="container">
          {isLoading ? (
            <div className="loadingPage">
              <div className="loadingInfo">
                <h1>LOADING....</h1>
              </div>
            </div>
          ) : (
            <div>
              {!!(ads && ads.length) ? (
                <>
                  <div className="listContainer">
                    <div className="contaienrTittle">
                      <h1>ADVERTISEMENTS AVIABLE</h1>
                    </div>
                    <ul>
                      {advertsToDisplay
                        .sort((a, b) => a.createdAt > b.createdAt)
                        .map(advert => (
                          <li key={advert.id}>
                            <div className="advert-container">
                              <Link to={`/adverts/${advert.id}`}>
                                <Advert
                                  advert={advert}
                                  onImageError={handleImageError}
                                />
                              </Link>
                            </div>
                          </li>
                        ))}
                    </ul>
                  </div>
                  <div className="pagination">
                    <p>
                      <span
                        className={currentPage === 1 ? "disabled" : "page"}
                        onClick={() => handlePageChange(currentPage - 1)}>
                        &lt;{" "}
                      </span>
                      {[...Array(totalPages)].map((_, index) => {
                        if (
                          totalPages > 5 &&
                          index > 1 &&
                          index < totalPages - 2
                        ) {
                          return (
                            <span className="page" key={index}>
                              ...
                            </span>
                          );
                        } else {
                          return (
                            <>
                              <span
                                className="page"
                                key={index}
                                onClick={() => handlePageChange(index + 1)}>
                                {index + 1}
                              </span>
                              {index < totalPages - 1 && <span> - </span>}
                            </>
                          );
                        }
                      })}
                      <span
                        className={isLastPage ? "disabled" : "page"}
                        onClick={() => handlePageChange(currentPage + 1)}>
                        {" "}
                        &gt;
                      </span>
                    </p>
                  </div>
                </>
              ) : (
                <EmptyList />
              )}
            </div>
          )}
        </div>
      </>
    </Layout>
  );
};

// const mapStateToProps = state => ({
//   adverts: getAdverts(state),
//   //isLoading: isLoading(state),
//   ...getUi(state),
// });
// const mapDispatchToProps = {
//   onAdvertsLoaded: advertsList,
// };

// export default connect(mapStateToProps, mapDispatchToProps)(AdvertsListPage);

export default AdvertsListPage;
