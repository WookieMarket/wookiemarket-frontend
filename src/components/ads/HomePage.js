import Layout from '../layout/Layout';
import { useTranslation } from 'react-i18next';
import AdvertsListPage from './AdvertsListPage';

const HomePage = (props) => {
  const { t } = useTranslation();
  return (
    <Layout title={t('welcome')} {...props}>
      <AdvertsListPage />
    </Layout>
  );
};

export default HomePage;
