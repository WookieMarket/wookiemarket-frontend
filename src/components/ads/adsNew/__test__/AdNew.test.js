/*import AdNew from "../AdNew";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import * as defaultState from "../../../../store/reducers";
import userEvent from "@testing-library/user-event";
import { adsCreate } from "../../../../store/slices/ads";

jest.mock("../../../../store/slices/ads", () => ({
  __esModule: true,
  adsCreate: jest.fn(), // Mock only the adsCreate action
}));

describe("AdNew", () => {
  const renderComponent = (error = null) => {
    const store = {
      getState: () => {
        const state = { ...defaultState }; // Create a copy of the state
        state.ui.error = error; // Modify the copy instead of the original state

        return state;
      },
      subscribe: () => {},
      dispatch: () => {},
    };
    return render(
      <Provider store={store}>
        <MemoryRouter>
          <AdNew />
        </MemoryRouter>
      </Provider>,
    );
  };

  test("snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("shoul dispatch adsCreate action", () => {
    const file = new File(["hello"], "hello.png", { type: "image/png" });

    const ad = {
      name: "mando",
      onSale: "true",
      price: "20",
      category: "imperio",
      description: "lado oscuro",
      status: "true",
      coin: "euro",
      image: file,
    };
    //NOTE I render the component
    renderComponent();

    const nameInput = screen.getByLabelText(/Article/);
    const onsaleInput = screen.getByLabelText(/Sell/);
    const priceInput = screen.getByLabelText(/Price/);
    const categoryInput = screen.getByLabelText(/Category/);
    const descriptionInput = screen.getByLabelText(/Description/);
    const statusInput = screen.getByLabelText(/Status/);
    const coinInput = screen.getByLabelText(/Coin/);
    const imageInput = screen.getByLabelText(/Image/);

    const submitButton = screen.getByTestId("buttonAdNew");
    expect(submitButton).toBeDisabled();

    //NOTE to launch events
    userEvent.type(nameInput, ad.name);
    userEvent.type(onsaleInput, ad.onSale);
    userEvent.type(priceInput, ad.price);
    userEvent.type(categoryInput, ad.category);
    userEvent.type(descriptionInput, ad.description);
    userEvent.type(statusInput, ad.status);
    userEvent.type(coinInput, ad.coin);
    userEvent.upload(imageInput, ad.image);

    expect(submitButton).toBeEnabled();
    userEvent.click(submitButton);

    expect(imageInput.files[0]).toStrictEqual(file);
    expect(imageInput.files.item(0)).toStrictEqual(file);
    expect(imageInput.files).toHaveLength(1);
    expect(adsCreate).toHaveBeenCalledWith(ad);
  });

  test("should display an error", () => {
    //NOTE // Spy on resetError function
    const resetErrorSpy = jest.spyOn(
      require("../../../../store/slices/ui"),
      "resetError",
    );

    const error = { message: "Network Error" };

    renderComponent(error);
    const errorElement = screen.getByText(error.message);

    expect(errorElement).toBeInTheDocument();
    const modalButton = screen.getByTestId("modalButton");
    userEvent.click(modalButton);

    expect(resetErrorSpy).toHaveBeenCalled();
  });
});*/
