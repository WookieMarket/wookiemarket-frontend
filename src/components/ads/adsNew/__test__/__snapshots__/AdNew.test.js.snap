// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdNew snapshot 1`] = `
<div>
  <div>
    <header
      class="header"
    >
      <nav
        class="navbar"
      >
        <div
          class="icons"
        >
          <a
            href="/"
          >
            <h1
              class="title"
            >
               
              header
               
            </h1>
          </a>
          <img
            alt="ES"
            class="icon-language"
            src="/images/languageIcons/es.png"
            title="icon-hover"
          />
          <img
            alt="EN"
            class="icon-language"
            src="/images/languageIcons/en.png"
            title="icon-hover"
          />
        </div>
        <ul>
          <li>
            <a
              class="navbar-list-item"
              href="/create-ad"
            >
              Upload ad
            </a>
            <a
              class="navbar-list-item"
              href="/ads"
            >
              Ads
            </a>
            <button
              class="navbar-list-item"
            >
              Logout
            </button>
          </li>
        </ul>
      </nav>
      <hr />
    </header>
    <main
      class="content"
    >
      <h2
        class="layout-title"
      >
        Create an ad
      </h2>
      <form
        class="container-form-creation"
        enctype="multipart/form-data"
      >
        <div
          class="form-group-creation"
        >
          <label
            class="password-label-creation"
            for="adname"
          >
            Article
          </label>
          <input
            class="password-input-creation"
            id="adname"
            name="name"
            placeholder="Name"
            type="text"
            value=""
          />
        </div>
        <div
          class="input-checked"
        >
          <div>
            <label
              class="checked-label-creation"
              for="onsale"
            >
              Sell
            </label>
            <input
              class="checked-input-creation"
              id="onsale"
              name="onSale"
              type="radio"
              value="true"
            />
          </div>
          <div>
            <label
              class="checked-label-creation"
              for="onsale"
            >
              Buy
            </label>
            <input
              class="checked-input-creation"
              id="onsale"
              name="onSale"
              type="radio"
              value="false"
            />
          </div>
        </div>
        <div
          class="form-group-creation"
        >
          <label
            class="password-label-creation"
            for="price"
          >
            Price
          </label>
          <input
            class="password-input-creation"
            id="price"
            name="price"
            placeholder="Price"
            type="number"
            value=""
          />
        </div>
        <div
          class="form-group-creation"
        >
          <label
            class="password-label-creation"
            for="category"
          >
            Category
          </label>
          <input
            class="password-input-creation"
            id="category"
            name="category"
            placeholder="Category"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group-creation"
        >
          <label
            class="password-label-creation"
            for="description"
          >
            Description
          </label>
          <input
            class="password-input-creation"
            id="description"
            name="description"
            placeholder="Description"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group-creation"
        >
          <label
            class="password-label-creation"
            for="status"
          >
            Status
          </label>
          <input
            class="password-input-creation"
            id="status"
            name="status"
            placeholder="Status"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group-creation"
        >
          <label
            class="password-label-creation"
            for="coin"
          >
            Coin
          </label>
          <input
            class="password-input-creation"
            id="coin"
            name="coin"
            placeholder="Coin"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group-creation"
        >
          <label
            class="password-label-creation"
            for="img"
          >
            Image
          </label>
          <input
            accept="image/*"
            class="img-input-creation"
            id="img"
            name="image"
            type="file"
            value=""
          />
        </div>
        <button
          class="sc-beqWaB dAsVMc"
          data-testid="buttonAdNew"
          disabled=""
          type="submit"
          width="button-form"
        >
          Create
        </button>
      </form>
    </main>
    <footer>
      <hr />
      <div
        class="footer"
      >
        <p>
          footer
        </p>
      </div>
    </footer>
  </div>
</div>
`;
